### PR TITLE
Allow applying `CompactionFilter` outside of compaction (facebook#8243)

### PR DESCRIPTION
**Pull request by MrCroxx**
Cherry-picking facebook/rocksdb#8243

> From HISTORY.md release note:
> 
> * Allow `CompactionFilter`s to apply in more table file creation scenarios such as flush and recovery. For compatibility, `CompactionFilter`s by default apply during compaction. Users can customize this behavior by overriding `CompactionFilterFactory::ShouldFilterTableFileCreation()`.
> * Removed unused structure `CompactionFilterContext`
> 
> Test Plan: added unit tests

Signed-off-by: MrCroxx <mrcroxx@outlook.com>

### CODE REVIEW

**Comment by MrCroxx** [comment]
There are still some failed test. Working on them.

File: HISTORY.md [hunk 1]
```diff
@@ -8,6 +8,7 @@
 * When user uses options.force_consistency_check in RocksDb, instead of crashing the process, we now pass the error back to the users without killing the process.
 * Added experimental ColumnFamilyOptions::sst_partitioner_factory to define determine the partitioning of sst files. This helps compaction to split the files on interesting boundaries (key prefixes) to make propagation of sst files less write amplifying (covering the whole key space).
 * Option `max_background_flushes` can be set dynamically using DB::SetDBOptions().
+* Allowing compaction filter on flushing, picked from [facebook/rocksdb#pr8243](https://github.com/facebook/rocksdb/pull/8243).
 
 ### Bug Fixes
 * Fixed issue #6316 that can cause a corruption of the MANIFEST file in the middle when writing to it fails due to no disk space.
```
Comment (tabokie) [inline HISTORY.md:14]: Missing "* Removed unused structure `CompactionFilterContext`."?

**Comment by MrCroxx** [comment]
All tests for compaction filter passed. 

New tests:
- DBTestCompactionFilter.CompactionFilterFlush
- DBTestCompactionFilter.CompactionFilterRecovery

```
./db_compaction_filter_test
[==========] Running 18 tests from 2 test cases.
[----------] Global test environment set-up.
[----------] 13 tests from DBTestCompactionFilter
[ RUN      ] DBTestCompactionFilter.CompactionFilter
[       OK ] DBTestCompactionFilter.CompactionFilter (1238 ms)
[ RUN      ] DBTestCompactionFilter.CompactionFilterDeletesAll
[       OK ] DBTestCompactionFilter.CompactionFilterDeletesAll (138 ms)
[ RUN      ] DBTestCompactionFilter.CompactionFilterFlush
[       OK ] DBTestCompactionFilter.CompactionFilterFlush (88 ms)
[ RUN      ] DBTestCompactionFilter.CompactionFilterRecovery
[       OK ] DBTestCompactionFilter.CompactionFilterRecovery (84 ms)
[ RUN      ] DBTestCompactionFilter.CompactionFilterWithMergeOperator
[       OK ] DBTestCompactionFilter.CompactionFilterWithMergeOperator (210 ms)
[ RUN      ] DBTestCompactionFilter.CompactionFilterContextManual
[       OK ] DBTestCompactionFilter.CompactionFilterContextManual (90 ms)
[ RUN      ] DBTestCompactionFilter.CompactionFilterContextCfId
[       OK ] DBTestCompactionFilter.CompactionFilterContextCfId (132 ms)
[ RUN      ] DBTestCompactionFilter.CompactionFilterIgnoreSnapshot
[       OK ] DBTestCompactionFilter.CompactionFilterIgnoreSnapshot (141 ms)
[ RUN      ] DBTestCompactionFilter.SkipUntil
[       OK ] DBTestCompactionFilter.SkipUntil (115 ms)
[ RUN      ] DBTestCompactionFilter.SkipUntilWithBloomFilter
[       OK ] DBTestCompactionFilter.SkipUntilWithBloomFilter (83 ms)
[ RUN      ] DBTestCompactionFilter.IgnoreSnapshotsFalse
[       OK ] DBTestCompactionFilter.IgnoreSnapshotsFalse (1081 ms)
[ RUN      ] DBTestCompactionFilter.IgnoreSnapshotsFalseDuringFlush
[       OK ] DBTestCompactionFilter.IgnoreSnapshotsFalseDuringFlush (1064 ms)
[ RUN      ] DBTestCompactionFilter.IgnoreSnapshotsFalseRecovery
[       OK ] DBTestCompactionFilter.IgnoreSnapshotsFalseRecovery (71 ms)
[----------] 13 tests from DBTestCompactionFilter (4536 ms total)

[----------] 5 tests from DBTestCompactionFilterWithCompactOption/DBTestCompactionFilterWithCompactParam
[ RUN      ] DBTestCompactionFilterWithCompactOption/DBTestCompactionFilterWithCompactParam.CompactionFilterWithValueChange/0
[       OK ] DBTestCompactionFilterWithCompactOption/DBTestCompactionFilterWithCompactParam.CompactionFilterWithValueChange/0 (967 ms)
[ RUN      ] DBTestCompactionFilterWithCompactOption/DBTestCompactionFilterWithCompactParam.CompactionFilterWithValueChange/1
[       OK ] DBTestCompactionFilterWithCompactOption/DBTestCompactionFilterWithCompactParam.CompactionFilterWithValueChange/1 (916 ms)
[ RUN      ] DBTestCompactionFilterWithCompactOption/DBTestCompactionFilterWithCompactParam.CompactionFilterWithValueChange/2
[       OK ] DBTestCompactionFilterWithCompactOption/DBTestCompactionFilterWithCompactParam.CompactionFilterWithValueChange/2 (913 ms)
[ RUN      ] DBTestCompactionFilterWithCompactOption/DBTestCompactionFilterWithCompactParam.CompactionFilterWithValueChange/3
[       OK ] DBTestCompactionFilterWithCompactOption/DBTestCompactionFilterWithCompactParam.CompactionFilterWithValueChange/3 (1021 ms)
[ RUN      ] DBTestCompactionFilterWithCompactOption/DBTestCompactionFilterWithCompactParam.CompactionFilterWithValueChange/4
[       OK ] DBTestCompactionFilterWithCompactOption/DBTestCompactionFilterWithCompactParam.CompactionFilterWithValueChange/4 (937 ms)
[----------] 5 tests from DBTestCompactionFilterWithCompactOption/DBTestCompactionFilterWithCompactParam (4754 ms total)

[----------] Global test environment tear-down
[==========] 18 tests from 2 test cases ran. (9291 ms total)
[  PASSED  ] 18 tests.
```

**Comment by MrCroxx** [comment]
[tikv/rust-rocksdb#pr635](https://github.com/tikv/rust-rocksdb/pull/635) - make `rust-rocksdb` expose the interface.

File: HISTORY.md [hunk 2]
```diff
@@ -8,6 +8,8 @@
 * When user uses options.force_consistency_check in RocksDb, instead of crashing the process, we now pass the error back to the users without killing the process.
 * Added experimental ColumnFamilyOptions::sst_partitioner_factory to define determine the partitioning of sst files. This helps compaction to split the files on interesting boundaries (key prefixes) to make propagation of sst files less write amplifying (covering the whole key space).
 * Option `max_background_flushes` can be set dynamically using DB::SetDBOptions().
+* Allowing compaction filter on flushing, picked from [facebook/rocksdb#pr8243](https://github.com/facebook/rocksdb/pull/8243).
+* Remove unused struct `CompactionFilterContext`.
```
Comment (tabokie) [inline HISTORY.md:12]: File it under `Public API change`, as original PR?

**Review by tabokie — APPROVED** [review]
Rest LG

File: HISTORY.md (hunk 2 above)
Comment (MrCroxx) [reply HISTORY.md]: Solved!

File: db/builder.cc [hunk 3]
```diff
@@ -215,8 +215,7 @@ Status BuildTable(
       meta->fd.file_size = file_size;
       meta->marked_for_compaction = builder->NeedCompact();
       assert(meta->fd.GetFileSize() > 0);
-      tp = builder
-               ->GetTableProperties();  // refresh now that builder is finished
+      tp = builder->GetTableProperties();  // refresh now that builder is finished
```
Comment (tabokie) [inline db/builder.cc:218]: BTW I think there is something wrong with your local diff-format, letter count should be capped beneath 80.

Comment (tabokie) [reply db/builder.cc:218]: FYI I'm using `git diff -U0 HEAD | clang-format-diff.py -p1 -i` with source file from https://github.com/llvm-mirror/clang/blob/master/tools/clang-format/clang-format-diff.py.

Comment (MrCroxx) [reply db/builder.cc:218]: 😨 Let me check it.

**Review by Connor1996 — APPROVED** [review]
LGTM

File: HISTORY.md [hunk 4]
```diff
@@ -3,6 +3,8 @@
 ### Public API Change
 * DeleteRange now returns `Status::InvalidArgument` if the range's end key comes before its start key according to the user comparator. Previously the behavior was undefined.
 * ldb now uses options.force_consistency_checks = true by default and "--disable_consistency_checks" is added to disable it.
+* Allowing compaction filter on flushing, picked from [facebook/rocksdb#pr8243](https://github.com/facebook/rocksdb/pull/8243).
```
Comment (tabokie) [inline HISTORY.md:6]: Why is this moved here😂, please double check the original PR.

Comment (MrCroxx) [reply HISTORY.md]: Sorry I think I misunderstood that. I fixed it referring to the origin PR. 😭
